add(doc): how-to Configure Ubuntu Core boot parameters (#39)
canonical/real-time-ubuntu-docs: docs/how-to/uc-boot-parameters.rst

How to Configure Boot parameters on Ubuntu Core 
================================================

Different from classic versions of Ubuntu (Desktop and Server), `Ubuntu Core`_ is a minimal, secure and strictly
confined operating system designed to run on embedded systems, such as IoT devices.
Ubuntu Core is a snap-based operating system, meaning it uses snap packages to manage applications and services.
Its read-only nature makes it a very secure and reliable operating system but also makes it a bit more difficult to
configure certain aspects of the system, such as boot parameters.

> 

Reviewer: Explain that this how-to assumes the use of a pre-built image: https://ubuntu.com/core/docs/try-pre-built-images
And that it is necessary to do it on a device and not a VM (for the same reasons specified in other docs).

Configure boot parameters
-------------------------

As UC doesn't have the file :code:`/etc/default/grub` mapped for the kernel parameters, neither the `update-grub`_
utility, it's necessary to use the `snap set`_. There are two ways to set the kernel parameters:

- `system system.kernel.cmdline-append`_
- `system system.kernel.dangerous-cmdline-append`_

The first one is used to dynamically append permitted kernel boot parameters that are verified against an `allow list`_
in the `gadget snap`_, while the second one is used to append kernel boot parameters considered risky according to that
specific gadget snap. In the case that you're interested in using the :code:`pc-gadget` or the :code:`pi-gadget`, you
should use the `system system.kernel.dangerous-cmdline-append` option, since the allow list isn't present on those 

> gadget snaps, but if you want to `build a gadget snap`_ yourself, you can implement it.

Reviewer: ```suggestion
should set the kernel parameters via the `system.kernel.dangerous-cmdline-append` option.
This is because those reference gadgets are generic and do not bundle an allowed parameter list.

When creating `a custom gadget snap`_ for a real-time Ubuntu Core, it is advisable to define the list of kernel parameters that are relevant to the use cases.
```

> specific gadget snap. When using the :code:`pc-gadget` or the :code:`pi-gadget`, it is advisable to use the 
> `system.kernel.dangerous-cmdline-append` option, as the allow list is not present in those gadget snaps. 
> However, in the case of `building a gadget snap`_ independently, it is possible to implement the allow list.

Reviewer: The fixed char and justified text is making it hard to suggest on the right sentences.

```suggestion
specific gadget snap.

When using the trial images, you need to use the 
`system.kernel.dangerous-cmdline-append` option, as the allow list is not present in the reference gadget snaps. 

When creating a custom `gadget snap`_ for a real-time Ubuntu Core system, it is advisable to add an allow list with the relevant parameters.
```

I think maybe we shouldn't even mention `system system.kernel.cmdline-append` and so avoid having to explain it. 

These and the custom gadget below to another how-to.

Author: We could omit this, but this is a small detail to explain (two sentences), I don't see much benefit on having this information missing. 
We would already have to explain what this command does anyway, if we only explain the `dangerous` version, it would give the wrong idea that it is the only way to do it, so it would be necessary to at least refer it anyway.

The kernel parameters in the snap configurations exactly match the `kernel parameters`_ that would be passed to the 
kernel on the grub command line, the :code:`GRUB_CMDLINE_LINUX_DEFAULT` on :code:`/etc/default/grub` file. 
Parameters are passed as a single string in the :code:`key=value` format, with each parameter separated by a space.

In the case of real-time kernel parameters, the most interesting parameters are `nohz`_, `nohz_full`_ and `irqaffinity`_.
The :code:`nohz` is used to enable/disable dynamic ticks (possible values are `on` or `off`), the :code:`nohz_full` 

> receives a `cpu list`_ where you say in which CPUs the dynamic ticks will be disabled, and the `irqaffinity` is used to 

Reviewer: ```suggestion
receives a `cpu list`_ where you specify in which CPUs the dynamic ticks will be disabled, and the `irqaffinity` is used to 
```

> set the affinity of the IRQs to the CPUs, in other words: which CPUs will handle the IRQs.

Reviewer: Too long. Turn to bullets?

> .. code-block:: bash

Reviewer: ```suggestion
For example:
.. code-block:: bash
```

    
    snap set system system.kernel.dangerous-cmdline-append="nohz=on nohz_full=2-N irqaffinity=0-1"

After setting the kernel parameters, it's necessary to reboot the system to apply the changes.
Then you can check if the parameters were applied:

.. code-block:: console

    $ cat /proc/cmdline
    snapd_recovery_mode=run console=ttyS0,115200n8 console=tty1 panic=-1 nohz=on nohz_full=2-N irqaffinity=0-1

> You may wish to undo the modifications made to the kernel parameters. To do so, you can use the `snap unset`_ command:

Reviewer: ```suggestion
To undo the modifications made to the kernel parameters, use the `snap unset`_ command:
```

> How to Configure Boot parameters on Ubuntu Core 

Reviewer: ```suggestion
How to configure boot parameters on Ubuntu Core 
```

================================================

Different from classic versions of Ubuntu (Desktop and Server), `Ubuntu Core`_ is a minimal, secure and strictly confined operating system designed to run on embedded systems, such as IoT devices.
Ubuntu Core is a snap-based operating system, meaning it uses snap packages to manage applications and services.
Its read-only nature makes it a very secure and reliable operating system but also makes it a bit more difficult to configure certain aspects of the system, such as boot parameters.

This how-to guide assumes the use of a `pre-built Ubuntu Core image`_ on a physical device rather than a virtual machine.
While a virtual machine can be used to follow the steps in this guide, it will not provide the benefits of a real-time kernel.

Configure boot parameters
-------------------------

As UC doesn't have the file :code:`/etc/default/grub` mapped for the kernel parameters, neither the `update-grub`_ utility, it's necessary to use the `snap set`_.
There are two ways to set the kernel parameters:

- `system system.kernel.cmdline-append`_

> - `system system.kernel.dangerous-cmdline-append`_

Author: Done, thank you!

The first one is used to dynamically append permitted kernel boot parameters that are verified against an `allow list`_ in the `gadget snap`_. 

> The second one is used to append kernel boot parameters considered risky according to that specific gadget snap.

Author: `unvetted` brings spell check problems, changing for "not analyzed"

When using the `trial images`_, you need to use the `system.kernel.dangerous-cmdline-append` option, as the allow list is not present in the reference gadget snaps (`pc-gadget`_ and `pi-gadget`_).  

When creating a custom `gadget snap`_ for a real-time Ubuntu Core system, it is advisable to add an allow list with the relevant parameters.

The kernel parameters in the snap configurations exactly match the `kernel parameters`_ that would be passed to the kernel on the grub command line, the :code:`GRUB_CMDLINE_LINUX_DEFAULT` on :code:`/etc/default/grub` file. 
Parameters are passed as a single string in the :code:`key=value` format, with each parameter separated by a space.

In the case of real-time kernel parameters, the most interesting parameters are `nohz`_, `nohz_full`_ and `irqaffinity`_.
The :code:`nohz` is used to enable/disable dynamic ticks (possible values are `on` or `off`). 
The :code:`nohz_full` receives a `cpu list`_ specifying which CPUs will have dynamic ticks disabled. 

> The `irqaffinity` is used to set the affinity of the IRQs to the CPUs, in other words: which CPUs will handle the IRQs.

Reviewer: Add reference to /reference/kernel-boot-parameters.
